### PR TITLE
Dave carroll7/MTW contract with refund

### PR DESCRIPTION
**Pull request by dv3-000**
### Description

Expands MetaTransactionWallet contract to allow for meta-transaction submitters to get refunded for gas payments.

### Other changes

Preliminary/unfinished changes to metatransactionwallet.ts (MTW contract test file)

### Tested

Not tested

### Backwards compatibility

No pre-existing functions were altered in the MTW contract, only new ones were created to add refunding capabilities.

### Documentation

### CODE REVIEW

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 1]
```diff
@@ -167,6 +180,41 @@ contract MetaTransactionWallet is
       );
   }
 
+  /**
+   * @notice Returns the struct hash of the refundable meta-transaction
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param _nonce The nonce for this meta-transaction local to this wallet.
```
Comment (anna-carroll) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol:202]: _nit_: why is `_nonce` prefixed with an underscore but not the other params?

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 2]
```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
   }
 
   /**
-   * @notice Executes a transaction on behalf of the signer.`
+  * relayer? TODO determine if we can predict this request (question for komenci folks)
```
Comment (anna-carroll) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: remove / address TODOs throughout PR

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 3]
```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
...
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit == gasLeft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
+    require(address(this).balance >= gasLimit.mul(tx.gasprice).add(value), "insufficient balance");
+    require(msg.sender == tx.origin, "relayer must be EOA");
+    require(gasLimit <= block.gaslimit, "Impossible gas limit");
+    require(metaGasLimit < gasLimit, "metaGasLimit must be less than gasLimit");
+    if (data.length > 0) require(Address.isContract(destination), "Invalid contract address");
```
Comment (anna-carroll) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: hmm i wonder if you should revert on this. if they are calling a non-contract address with data, it almost certainly means they have submitted a malformed meta-transaction, but I wonder whether it would be better to remain un-opinionated about the contents of the meta-transaction - let users execute a messed up tx if they sign it & pay for it ¯\_(ツ)_/¯

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 4]
```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
...
+      metaGasLimit,
+      v,
+      r,
+      s
+    );
+    require(_signer == signer, "Invalid meta-transaction signer");
+    nonce = nonce.add(1);
+
+    // TODO: ask Contracts
+    uint256 buffer1 = 4949; // TODO: determine this constant (gas required for operations after destination.call.value)
```
Comment (anna-carroll) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: _nit_: rename something more informative - `postCallGasBuffer1` & `postCallGasBuffer2`?

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 5]
```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
...
+      destination,
+      value,
+      data,
+      nonce.sub(1),
+      maxGasPrice,
+      metaGasLimit,
+      returnData
+    );
+
+    msg.sender.transfer(gasLimit.sub(gasLeft()).add(buffer2).mul(tx.gasprice));
```
Comment (anna-carroll) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: hard for me to parse whether this calculation is correct. subtracting the gas left from the gas limit seems wrong to me but I don't have enough context.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 6]
```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
...
+      s
+    );
+    require(_signer == signer, "Invalid meta-transaction signer");
+    nonce = nonce.add(1);
+
+    // TODO: ask Contracts
+    uint256 buffer1 = 4949; // TODO: determine this constant (gas required for operations after destination.call.value)
+    uint256 buffer2 = 4747; // TODO: determine this constant (gas required for operations after and including msg.sender.transfer)
+
+    if (address(this).balance >= metaGasLimit.add(buffer1).mul(tx.gasprice).add(value)) {
```
Comment (anna-carroll) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: hmmm okay so if there's not enough gas to execute the external call then we just won't do the call, but we will increment the nonce? 🤔 this seems kinda fishy to me but again i don't have perfect context

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 5 above)
Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: i would do the calculations in a separate line from the `transfer` call and document them with comments

Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: also, it's generally best practice to use `.call()` with a `value` rather than using Solidity's built-in `transfer` - though, since you require that `msg.sender` is an EOA, it's less dire
https://consensys.net/diligence/blog/2019/09/stop-using-soliditys-transfer-now/

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 6 above)
Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: i guess if the reasoning is, if a user signed a meta-tx that's impossible to execute because they didn't provide a high enough gas limit, then we would want to be able to execute the transaction anyway and refund Komenci for doing so?

i would evaluate this statement on a separate line and document the reasoning with comments

Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: also: seems like the result of this statement should be included in the event - like `bool metaTransactionAttempted` so that it's easier to see off-chain whether the meta-tx execution was skipped

**Review by anna-carroll — COMMENTED** [review]
cool! 🌝

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 7]
```diff
@@ -33,12 +33,15 @@ contract MetaTransactionWallet is
   event WalletRecovered(address indexed newSigner);
   event EIP712DomainSeparatorSet(bytes32 eip712DomainSeparator);
   event Deposit(address indexed sender, uint256 value);
+  event FailedMetaTransaction(string error);
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: Nit: by convention, event names are usually "verby" rather than "nouny" - I'd reword to MetaTransactionFailed.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 8]
```diff
@@ -167,6 +180,41 @@ contract MetaTransactionWallet is
...
+    bytes memory data,
+    uint256 _nonce,
+    uint256 maxGasPrice,
+    uint256 gasLimit,
+    uint256 metaGasLimit
+  ) internal view returns (bytes32) {
+    return
+      keccak256(
+        abi.encode(
+          EIP712_EXECUTE_META_TRANSACTION_TYPEHASH,
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: Don't we need a new EIP712 typehash if these messages have a different structure?

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 9]
```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
...
+    uint256 metaGasLimit,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit == gasLeft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
+    require(address(this).balance >= gasLimit.mul(tx.gasprice).add(value), "insufficient balance");
+    require(msg.sender == tx.origin, "relayer must be EOA");
+    require(gasLimit <= block.gaslimit, "Impossible gas limit");
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: Why is this check necessary?

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 3 above)
Comment (m-chrzan) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: Agreed, we don't require this in `executeMetaTransaction`.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 4 above)
Comment (m-chrzan) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: Even better if the `1` and `2` can be replaced with something descriptive.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 10]
```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
...
+    require(metaGasLimit < gasLimit, "metaGasLimit must be less than gasLimit");
+    if (data.length > 0) require(Address.isContract(destination), "Invalid contract address");
+
+    // checking gasCurrency is out of scope, would require pre-compile
```
Comment (m-chrzan) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol:417]: Haven't thought through this all the way, but I'm wondering if we maybe could implement specifying a gas currency without the need for an on-chain check here? As long as the MTW owner specifies the currency they will refund with, the gasLimit and gasPrice, they won't get charged more than they expected, so it doesn't really matter what currency the relayer actually used.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 9 above)
Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: +1

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 4 above)
Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: ^^ +1

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 6 above)
Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:411]:
```
// determine whether there is enough gas left 
// to execute the meta transaction
uint256 gasNeeded = metaGasLimit.add(buffer1);
bool enoughGas = gasLeft() >= gasNeeded;
// determine whether balance of wallet is enough
// to execute the meta transaction
uint256 currencyNeeded = gasNeeded.mul(tx.gasPrice).add(value);
bool enoughCurrency = address(this).balance >= currencyNeeded
// if enough gas & enough balance, execute transaction
if(enoughGas && enoughCurrency) {
```

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 4 above)
Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: maybe
```
buffer1 = gasNeededAfterMetaTransaction
buffer2 = gasNeededToTransferAndReturn
```

Comment (anna-carroll) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: define `buffer2` / `gasNeededToTransferAndReturn` where it's used, not here

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 11]
```diff
@@ -47,6 +53,16 @@ contract MetaTransactionWallet is
...
   );
 
+  event MetaTransactionWithRefundExecution(
+    address indexed destination,
+    uint256 value,
+    bytes data,
+    uint256 indexed nonce,
+    uint256 maxGasPrice,
+    uint256 metaGasLimit,
+    bytes returnData
```
Comment (alecps) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: One idea to simplify the events would be to just add a boolean success field here and get rid of the FailedMetaTransaction event

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 12]
```diff
@@ -236,7 +348,102 @@ contract MetaTransactionWallet is
...
+
+    bytes memory returnData;
+
+    {
+      // TODO: ask Contracts
+      uint256 buffer1 = 4949; // TODO: determine this constant (gas required for operations after destination.call.value)
+
+      if (
+        address(this).balance >= metaGasLimit.add(buffer1).mul(tx.gasprice).add(value) &&
+        metaGasLimit < (gasleft() * 63) / 64
```
Comment (alecps) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]:
```suggestion
        metaGasLimit <= gasleft() * 63/64
```

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 13]
```diff
@@ -236,7 +365,101 @@ contract MetaTransactionWallet is
...
+    uint256 maxGasPrice,
+    uint256 gasLimit,
+    uint256 metaGasLimit,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit >= gasleft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
+    require(address(this).balance >= gasLimit.mul(tx.gasprice).add(value), "insufficient balance");
```
Comment (nategraf) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol:413]: If at all possible, we should support all whitelisted fee tokens (e.g. cUSD and cEUR). If we don't, this will be much less useful to average Celo wallet users who hold only stable tokens.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 14]
```diff
@@ -236,7 +365,101 @@ contract MetaTransactionWallet is
...
+    require(msg.sender == tx.origin, "relayer must be EOA");
```
Comment (nategraf) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol:414]: What is the concern with non-EOA relayers? I could imagine a legitimate uses for contract relayers, such as bundling multiple relayed transactions together to increase fee efficiency. If there is an attack we are trying to prevent, it would be good to document that here.

One more note here is that EIP-3074 will break this check if deployed to Celo. We may need to keep this in mind to try to be forward-compatible.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 15]
```diff
@@ -236,7 +365,101 @@ contract MetaTransactionWallet is
...
+    bytes calldata data,
+    uint256 maxGasPrice,
+    uint256 gasLimit,
+    uint256 metaGasLimit,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit >= gasleft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
```
Comment (nategraf) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: An alternative to this would be specify they refunded gas price as `min(tx.gasPrice, maxGasPrice)`. The advantage would be to allow for relayers to use a higher gas price to get a transaction cleared, but if they do, they will only be refunded up to the amount the user agreed to ahead of time.

Comment (nategraf) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol]: If I am not mistaken, we could do the same for the overall gas limit. Assuming the user is not intending to cause an out-of-gas error (which is not impossible, but probably not something we are intending to support through this mechanism), this should not violate the user's wishes if a relayer wants to pay for some extra gas.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 10 above)
Comment (nategraf) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:417]: +1 to this. The parameters set by the user can be seen as an agreement from the user to pay the relayer at some rate per unit of gas used, and this could be in any token, even those that are not whitelisted. If the relayer then wants to pay for the transaction gas fees in a different currency, that should be perfectly fine.

Comment (nategraf) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:417]: (In this case we may not be able to easily compare the gas price on the transaction overall, and that of the refund because they may be different currencies. Two options to address this would be to either use the maxGasPrice specified by the user as the refund rate, or add a relayer specified refund gas price parameter that can be set up to the max gas price. (Downside to both of these is that the relayer can arbitrage any difference between the user-specified max gas price and the one they cleared the transaction at when they are in different currencies). Another option would be to actually check the on-chain oracle for the exchange rate to compare gas prices.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 16]
```diff
@@ -236,7 +365,101 @@ contract MetaTransactionWallet is
...
+        address(this).balance >=
+        metaGasLimit.add(gasAfterMTxExecution).mul(tx.gasprice).add(value) &&
+        metaGasLimit < (gasleft() * 63) / 64
+      ) {
+        bool success;
+        {
+          // if value is non-zero, destination.call seems to add 2300 gas to metaGasLimit.
+          // We want this to execute with exactly the metaGasLimit specified by the signer
+          // so we subtract 2300 from metaGasLimit if value is non-zero
+          if (value > 0) metaGasLimit = metaGasLimit.sub(2300);
```
Comment (nategraf) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol:445]: it seems off to modify this variable. E.g the event below `MetaTransactionWithRefundExecution` would now output this lower value. It would be better to use a new local variable to store the modified amount. (i.e. to treat the local variable as immutable)

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 17]
```diff
@@ -236,7 +365,101 @@ contract MetaTransactionWallet is
...
+    nonce = nonce.add(1);
+
+    bytes memory returnData;
+
+    {
+      // TODO: ask Contracts about a more accurate estimate
+      uint256 gasAfterMTxExecution = 7900; // estimated by emitting gasleft() at desired benchmark times
+
+      if (
+        address(this).balance >=
```
Comment (nategraf) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol]: If I understand correctly, nothing could have happened since the balance check at the start of this function that would change the contract balance. If this is true, this balance check feels redundant.

File: packages/protocol/contracts/common/MetaTransactionWallet.sol [hunk 18]
```diff
@@ -236,7 +365,101 @@ contract MetaTransactionWallet is
...
 
   /**
-   * @notice Executes a transaction on behalf of the signer.`
+   * @notice Executes a refundable meta-transaction on behalf of the signer.
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param maxGasPrice The maximum gas price the user is willing to pay.
+   * @param gasLimit The gas limit for entire relayed transaction.
+   * @param metaGasLimit The gas limit for just the meta-transaction.
```
Comment (nategraf) [inline packages/protocol/contracts/common/MetaTransactionWallet.sol:393]: It is a bit unclear to me what the (user-signed) `metaGasLimit` is designed to protect against. My guess is that we need a way to assure relayers there is enough gas left-over to pay for the refund, and we need to the user to sign it because we wouldn't want the relayer to set it arbitrarily low and cause the user transaction to revert while still receiving a refund. Is that accurate? If so it would be helpful to clarify that.

Also, if that is true, and alternative mechanism to this would be to have a fixed constant amount of gas reserved for excuting the checks and refund logic. Was this considered?

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 16 above)
Comment (dv3-000) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:445]: Oddly the `destination.call` execution actually mutates `metaGasLimit` outside of the call itself, so if we don't subtract 2300 from `metaGasLimit`, once it is emitted it's actually a larger value than expected

Comment (dv3-000) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:445]: Even more odd is when I assign a new local variable to store the value of `metaGasLimit` and pass that new local variable instead e.g. `uint256 _metaGasLimit = metaGasLimit; if (value > 0) _metaGasLimit = _metaGasLimit.sub(2300); (success, returnData) = destination.call.value(value).gas(_metaGasLimit)(data);` the values of both `_metaGasLimit` and `metaGasLimit` are both mutated by `destination.call`

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 18 above)
Comment (alecps) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:393]: Yea that's exactly right. I suppose we could do away with the metaGasLimit like you suggest and it would simplify the code. My concern is that I think the gas requirements can vary based on state, though I'm not sure by how much. Do you have any insight into this?

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 14 above)
Comment (alecps) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:414]: Agreed we probably don't need this check, we were just trying to simplify our set of assumptions and had no particular attack in mind

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 13 above)
Comment (alecps) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:413]: Agreed, though it may end up being out of scope for this project

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 10 above)
Comment (alecps) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:417]: If I can try to summarize:
- No matter what, the signer should specify the refund currency, and (max)gasPrice/gasLimit denominated in that currency. 
- If the relayer wants to pay gas in a different currency, that's okay as long as we can ensure they are refunded as specified by the signer. 
- If the relayer chooses to pay in a different currency, then tx.gasprice (specified by the relayer) will not be denominated in the same currency as the signer specified (max)gasPrice. 
- If the relayer uses a gas currency that is more valuable on a unit basis, that is fine since tx.gasprice < (max)gasPrice. They can do this with the implementation we have already. 
- If we want to allow the relayer to pay for gas in a currency that is less valuable on a unit basis, then we will need to use an oracle to convert the gas prices into the same units for comparison.
- In order to use an oracle, we need to know what currency the relayer is paying for gas in because otherwise they could claim to use a less valuable currency than they actually submit the transaction with, and get refunded at a higher rate than they should be. 
- We cannot know what gas currency the relayer actually submits the transaction with because it is not exposed to the smart contracts.
- Unless I've made an error above (I may well have) I think this means we can't support this feature without exposing tx.gascurrency. 

Does this reasoning check out?

File: packages/protocol/contracts/common/MetaTransactionWallet.sol (hunk 16 above)
Comment (alecps) [reply packages/protocol/contracts/common/MetaTransactionWallet.sol:445]: This is probably due to some strange solidity storage stuff. Let's investigate this more together @dave-carroll7
